### PR TITLE
RFC: feat(awful: hotkeys_popup): add possibility to create new widget…

### PR DESCRIPTION
**Pull request by actionless**
… instance

### CODE REVIEW

**Comment by Elv13** [comment]
I am fine with this. Finish the WIP, add some documentation and I will merge it.

One of my next PR will allow widgets to be converted to svg/png/surface. This will allow the awesome help to be included in the wallpaper itself (if the theme enable it). So this commit is  kind of necessary to make it work.

/offtopic One feature my old implementation has and this one doesn't have is the ability to replace groups by custom widgets.

**Comment by actionless** [comment]
>/offtopic One feature my old implementation has and this one doesn't have is the ability to replace groups by custom widgets.

i like the idea and that should address initial @blueyed 's proposal

**Comment by actionless** [comment]
do you have any ideas regarding my question above?

also, i am not sure how it will be better to distinguish in document methods of the module itself and the widget instance created by it

**Comment by actionless** [comment]
@blueyed i remember you were the original requester, please have a look

File: lib/awful/hotkeys_popup/widget.lua [hunk 1]
```diff
@@ -96,351 +46,438 @@ widget.hide_without_description = true
...
+                    widget_instance.additional_hotkeys
+                    )
+                end
             end
         end
-    end)
+        sort_hotkeys(widget_instance.additional_hotkeys)
+    end
+
+
```
Comment (blueyed) [inline lib/awful/hotkeys_popup/widget.lua]: Too much whitespace.. ;)

**Comment by actionless** [comment]
@blueyed done

**Comment by blueyed** [comment]
I've seen this error when trying to use the "hotkeys" menu entry from the default config.
Might not be relevant to this PR, but the hotkeys-popup feature in general though.

```
2016-11-20 22:56:38 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:311: in upvalue 'create_wibox'
	/usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:406: in function 'awful.hotkeys_popup.show_help'
	/usr/share/awesome/lib/awful/menu.lua:311: in function 'awful.menu.exec'
	/usr/share/awesome/lib/awful/menu.lua:441: in upvalue 'press'
	/usr/share/awesome/lib/awful/button.lua:46: in function </usr/share/awesome/lib/awful/button.lua:46>
	[C]: in method 'emit_signal'
	/usr/share/awesome/lib/wibox/widget/base.lua:367: in function 'wibox.widget.base.handle_button'
	(...tail calls...)
	/usr/share/awesome/lib/gears/object.lua:116: in function 'gears.object.emit_signal'
	/usr/share/awesome/lib/wibox/drawable.lua:405: in function </usr/share/awesome/lib/wibox/drawable.lua:400>
error: /usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:311: attempt to get length of a nil value (field 'widgets')
```

It worked when trying it again though.

**Comment by blueyed** [comment]
To create a separate widget, I would have to use `widget_module.new()` (https://github.com/awesomeWM/awesome/blob/384f647d264af61bbff5ff1c909a0c192c3c4bea/lib/awful/hotkeys_popup/widget.lua#L48), no?
But it is not available on the module?!

**Comment by actionless** [comment]
it's available, i've just forgot to add docstrings to that function

**Comment by blueyed** [comment]
Tried to use it as `return require("awful.hotkeys_popup").widget.new()` in awesome-client - returns nil?!

**Comment by blueyed** [comment]
Oh sorry, I am apparently not using/testing this PR currently.. :/

**Comment by blueyed** [comment]
Hmm - I might totally miss something here, but `return require("awful.hotkeys_popup").widget.new` still returns `nil` for me..

**Comment by actionless** [comment]
thanks for having a look!

**Comment by psychon** [comment]
Travis broke: https://travis-ci.org/awesomeWM/awesome/builds/177638363

**Comment by actionless** [comment]
i've restrted it for now, will look further if it'll fail again

**Comment by actionless** [comment]
the same result, is it possible what version of luacheck changes?

**Comment by psychon** [comment]
The failing build https://travis-ci.org/awesomeWM/awesome/jobs/177638368 has luacheck 0.17.0-1.
Last working build https://travis-ci.org/awesomeWM/awesome/jobs/176522591 has luacheck 0.16.3-1.
Detecting trailing whitespace and inconsistent indentation is the new feature in luacheck 0.17.0: https://github.com/mpeterv/luacheck/releases/tag/0.17.0

Yup, the version changed.

**Comment by actionless** [comment]
oops, i've got updated my previous comment almost in the same time

**Comment by actionless** [comment]
i think fixing the warnings will be better than falling back to the previous release/chaning check config

**Comment by actionless** [comment]
i can do it tonight

**Comment by blueyed** [comment]
@actionless 
Fixed already.. :)

**Pull request merged**